core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java

```diff
@@ -386,22 +392,30 @@ public AddPageWork(Block block)
         public boolean process()
         {
             int positionCount = block.getPositionCount();
-            checkState(lastPosition < positionCount, "position count out of bound");
 
-            // needRehash() == false indicates we have reached capacity boundary and a rehash is needed.
             // We can only proceed if tryRehash() successfully did a rehash.
-            if (needRehash() && !tryRehash()) {
-                return false;
+            while (nextGroupId + positionCount >= maxFill) {
```

Comment:
Does that change make possible that some query fails now whileas that query does not fail before (when spill to disk is diabled)? 
I mean the situation when we have to process the last page and that page conatins 100k rows that are going to be assigned to one group.

Reply:
It was just an example. Obviously we do not need a such specific case to reveal that issue. Actually, the last page can contain much less rows - like 2k or 300 or less. And it is not required for the last page to contain one group. The last page can contain a lot of group as well and query still could fail (whileas it succeed before).

Reply:
I think that the requirements for that case are less probable than you think.
First of all in order to run out of memory query needs to be big.
So for a big enough query we have a case that:
- There is a lot of groups (so that the hash table is big enough). Let's assume 8M buckets (tpch/q18)
- The last page contains one or more groups that has not been seen before. Given that there already has been billions of incoming rows processed this is highly unlikely.
- The few added groups hit exactly the current limit. Given that the hash table is 4M buckets, there is a sort of 1/4M chance of that per any new group.
- Increasing the hash table size takes additional 4M*12B = 48MB. So now this number needs to hit exactly the memory limit. If the query is big enough to spawn millions of groups, the node/cluster probably has hundreds of GB of memory.

So, as you can see, the probability of hitting such an issue is minimal.
If you have a more likely example, please share it here.

Reply:
I can imagine that probability of such case is minimal. Rather, the question for me is: is that acceptable for in comparison to the performance gain from that. Very possibly, yes - we can accept it. It is kind of tradeoff (for me).